src/Xamarin.Android.Build.Tasks/Utilities/MetadataResolver.cs

```diff
@@ -0,0 +1,60 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+
+namespace Xamarin.Android.Tasks
+{
+	/// <summary>
+	/// A replacement for DirectoryAssemblyResolver, using System.Reflection.Metadata
+	/// </summary>
+	public class MetadataResolver : IDisposable
+	{
+		readonly Dictionary<string, PEReader> cache = new Dictionary<string, PEReader> ();
+		readonly List<string> searchDirectories = new List<string> ();
+
+		public MetadataReader GetAssemblyReader (string assemblyName)
+		{
+			var key = Path.GetFileNameWithoutExtension (assemblyName);
+			if (!cache.TryGetValue (key, out PEReader reader)) {
+				var assemblyPath = Resolve (assemblyName);
+				cache.Add (key, reader = new PEReader (File.OpenRead (assemblyPath)));
+			}
+			return reader.GetMetadataReader ();
+		}
+
+		public void AddSearchDirectory (string directory)
+		{
+			directory = Path.GetFullPath (directory);
+			if (!searchDirectories.Contains (directory))
+				searchDirectories.Add (directory);
+		}
+
+		public string Resolve (string assemblyName)
+		{
+			string assemblyPath = assemblyName;
+			if (!assemblyPath.EndsWith (".dll", StringComparison.OrdinalIgnoreCase)) {
+				assemblyPath += ".dll";
```

Comment:
Should this also support `.exe` suffixes?

Reply:
Since https://github.com/xamarin/xamarin-android/commit/e390702773365ffe3da975181d0aa398d42731d0 we dropped support for referencing exe's at the MSBuild level. I didn't add support for it here, but if there is some scenario you think we'll hit, I can add it.

Offhand I can't think of a case where a `.dll` would reference a random `.exe`?